### PR TITLE
Add DataSourcingActor

### LINKED ISSUE #30: Implement a data sourcing actor
### What's needed?

We need to implement an actor that can expose data from different sources to the downstream layers.

Component data can come from different sources, e.g., the Microgrid API, the Reporting API, and parquet files in local storage.
The data format in these sources can be different.

There may be mutiple downstream actors, each of them reading from multiple sources.
If they all have mechanisms to convert data into one common format for further consumption, then:
* they would each need to understand how all the different data sources work, causing high coupling, and
* they would very likely run similar conversions, wasting CPU cycles

### Proposed solution

The solution would be to create an actor that
* can read multiple data sources, e.g., the Microgrid API, and parquet files in local storage,
* can expose data in a common format that abstracts the underlying data source away,
* can switch between input data streams seamlessly, based upon well-defined conditions (see below for details).

Note that in the first version (v1), this actor needs to consume data from the following sources only:
* Microgrid API
* parquet files in local storage

In order for it to identify the data channel and the data source(s), the requests to the actor should consists of the following parameters:
a. the component ID
b. the metric ID
c. the start timestamp, or the timestamp from which data should be read

If the start timestamp is in past, then the local parquet files should be read first. Once there is no more data in the parquet files, then data from the live data source (Microgrid API) should be streamed.

The top-level building blocks of the actor could be a set of classes as follows
(Note that these are just pseudocoded ideas to specify requirements. The actual implementation may vary):

```py
# This is the class that uniquely represents a metric channel.
@dataclass
class ChannelKey:
    # An enum or a unique ID representing the metric.
    metric_id: str

    # The component ID
    component_id: int

    # The time from which data should be consumed.
    start_ts: datetime
```
See `ComponentMetricKey` in [issue #5](https://github.com/frequenz-floss/frequenz-sdk-python/issues/5) for more details on `ChannelKey`.

```py
# This is the class that consumes and forwards data for one given channel.
# The actor can have a collection of the objects of this class.
class DataIngestor:
  _channel: Option[BroadcastChannel[float]]

  def __init_(self):
    # Initialize `_channel` as `None`
    self._channel = None
    ...

  def get_channel(self, data_key: ChannelKey) -> BroadcastChannel[float]:
    # This method is key here.
    #
    # Consume data from different sources, and stream them downstream via an
    # output channel. The user has to input the details of which
    # metric has to be consumed and from what time. The user gets a channel with
    # the desired data in return.
    #
    # If `data_key.start_ts` is `now()`, then read directly from the
    # Microgrid API.
    #
    # If `data_key.start_ts` is in the past, e.g., `now() - 30m`, then read the
    # parquet files until there is no more data there. Then read from the
    # microgrid API.
    #
    # If `data_key.start_ts` is in the future, e.g., `now() + 30m`, then wait
    # until the given timestamp is reached. Then read from the microgrid API.

    # Return `self._channel` if it already exists, otherwise create a new one.
    ...
```

The actor itself could be defined as follows:

```py
@dataclass
class DataRequestKey:
  # The details of the metric to read data for, and the timesamo from which the
  # data should be read.
  data_key: ChannelKey

  # Can have more elements as required.

class DataSourcingActor:
  # The idea here is to store previously returned channels, so that we reuse them
  # insted of creating new ones every time. The data structure can be a `dict`, or
  # can be some other custom data structure, depending upon the following
  # requirements:
  #
  # Two `ChannelKey` objects with `start_ts >= now()` are
  # effectively the same, since they consume live data from the Microgrid API.
  #
  # However, two `ChannelKey` objects that need historical data from the
  # same `start_ts` timestamp may not be the same. Assuming the two
  # `ChannelKey` objects are received at times 00:00:00 and 01:00:00,
  # then the 2nd one will still require to start from the the first historical
  # data sample, whereas the 1st one could be several data samples ahead or
  # reading live data by then.
  _data_ingestors: Dict[DataRequestKey, DataIngestor]

  def __init__(self):
    # Initialize the `_data_ingestors` object.
    self._data_ingestors = {}
    ...

  def get_channel(self, data_key: ChannelKey):
    data_req_key = DataRequestKey(data_key)

    # If a `DataIngestor`1 object has not yet been created, then create it.
    if not self._data_ingestors.contains(data_req_key):
      self._data_ingestors[data_req_key] = DataIngestor()

    # Return the channel from the `DataNgestor` instance.
    return self._data_ingestors[data_req_key].get_channel()
```

### Use cases

_No response_

### Alternatives and workarounds

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by shsms**
Closes #30

### CODE REVIEW

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/common.py]: What about `messages.py` or `request.py`? I'm not crazy about modules called `common` because it's hard to figure out what could be inside (and with time they tend to have a tendency to be a mess full of unrelated stuff).

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/common.py]: I still don't fully understand why we want to construct the channel name like this if users are supposed to pass a UUID. If there is a UUID involved, then then channel name can simply be the UUID, it will be also known by both parties. Is all the other information in the channel name only used for debugging purposes?

At some point it was mentioned that `channel_uuid` was going to be named more like `channel_namespace`, which gives more the impression that 2 different users could use the same channel if they both use the same *namespace*, but if it is a UUID it looks like nobody will ever use the same channel name, even if subscribing to the same metric from the same component.

That said, I'm also OK with keeping it as it is until we gather more experience and a better view on how this will be used by the `*Pool` and `LogicalMeter` actors.

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/common.py]: I would name this `ComponentMetricId`, as we said that the name of field of a component is a [*Metric ID*](https://github.com/frequenz-floss/frequenz-sdk-python/wiki/Glossary#metric-id). It could also be simply `MetricId`, but I'm fine with making it more explicit that is a metric ID of components.

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/common.py]: I would also name this `metric_id`.

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: Why not `metric not in _BatteryDataMethods`? (and so on with all categories)?

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: I would name this file `micrigrid_api_source` too, to make it clear this is about API (live) data, as data coming from parquet files will also be *microgrid* data.

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: pro-tip: You can use [`dict.setdefault()`](_BatteryDataMethods) for this.

```suggestion
        self._req_streaming_metrics.setdefault(comp_id, {}).setdefault(request.metric, []).append(request)
```

Not sure if it makes it more clear or more opaque, though :)

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/common.py]: This is first required by the `MicrogridApiSource`, so I would move it directly to `microgrid_source.py`.

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/common.py]: This is first required by the `DataSourcingActor`, so I would move it to `data_sourcing.py`.

This way we can also completely get rid of the `common` module. Both `ComponentMetricRequest` and `ComponentMetricId` can be exposed to the user via `__init__`, so they can do `from frequenz.sdk.actor.data_sourcing import ComponentMetricRequest` for example, and they don't care where it is initially defined.

File: src/frequenz/sdk/actor/data_sourcing/data_sourcing.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/data_sourcing.py]: Why is the category resolution done at this level if it uses the microgrid API client? Shouldn't it be done by the `MicrogridApiSource` so the actor itself have no direct dependency to the microgrid API?

When getting data from parquet files at some point we'll need to check the category in some historical description of the microgrid too, right?

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: Good question.

I initially wrote a very long message with lots of different approaches with pros and cons, but I don't think it is really adding any value because I didn't come up with any solution I like better than what you are suggesting :laughing: 

So I would say let's go with `src/frequenz/sdk/data_pipeline/types.py` for now decide later if there is a better place.

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: As I said in the resampling PR, maybe we can just say `measurement` to be more consistent with the glossary. Also maybe mention why and when the value could be `None`:

```suggestion
    """A measurement taken at a particular point in time.
    
    The `value` could be `None` if a component is malfunctioning or data is lacking
    for another reason, but a sample still needs to be sent to have a coherent view
    on a group of component metrics for a particular timestamp.
    """
```

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]:
```suggestion
        """Create a `MicrogridApiSource` instance.
```

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: Nitpick: I would use a `ValueError` here, as `TypeError` are supposed to be mismatches on the type system.

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: This should be a `ValueError`.

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: This should be a `ValueError`.

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: Looking at `_update_strems()` a possible optimization here would be to check if the requested metric is already handled and avoid calling `_update_strems()` to avoid cancelling and re-creating the task.

**Review by leandro-lucarella-frequenz — COMMENTED** [review]
The microgrid source was a bit hard to follow, not sure if there is a simpler way to do it because I guess it has quite a bit of inherent complexity, but maybe more comments can be added. I feel I lacked a more high-level overview when reviewing about what are the responsibilities of each part, or the general algorithm, or even comments on the main attributes for microgrid source class.

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (sahas-subramanian-frequenz) [reply src/frequenz/sdk/actor/data_sourcing/common.py]: I was going to rename this to `namespace`, but forgot.  Will do in the next changeset.

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/common.py]: done

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/common.py]: Both `ComponentMetricId` and `ComponentMetricRequest` need to be accessible for other actors,  and will probably be accepted by the resampling actor too?  So I put it in the ~`frequenz/sdk/types.py`~ (EDIT: I meant `frequenz/sdk/data_pipeline/types.py`).  Does that sound ok?

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/common.py]: done

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/common.py]: done

File: src/frequenz/sdk/actor/data_sourcing/data_sourcing.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/data_sourcing.py]: I think the decision was that the parquet files will on the edge will get migrated when the component IDs change, etc.  So I think the microgrid api will be the source of this information for historical data as well.

Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/data_sourcing.py]: But I guess you're right,  the microgrid api won't be available when running the actor on the cloud to just read historical data.  I will move it into `MicrogridApiSource` tomorrow.

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: done

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: done

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: done

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: done

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: oh yes,  it is because I wrote these checks first,  before I wrote the data extraction methods.  Fixed now, thanks for noticing. :-)

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: done

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: done

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: done

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: it is more readable,  and people can look up the method,  updated, thanks. :-)

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: will do tomorrow.

**Review by shsms — COMMENTED** [review]
I will add some more comments tomorrow as well.

File: src/frequenz/sdk/actor/data_sourcing/microgrid_api_source.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/microgrid_api_source.py]: Nitpick: this is a bit redundant, `x in dict` is looking if `x` is in the keys of the dict already :)

```suggestion
        for metric in requests:
```

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (leandro-lucarella-frequenz) [reply src/frequenz/sdk/actor/data_sourcing/common.py]: For me it doesn't sound like a general interface, I see it more owned by the data sourcing actor, even if other actors use it. It is completely coupled with how the data sourcing actor works and will have to be updated if the data sourcing actor *interface* changes, this is why I think they belong together.

Yes, probably the resampling actor will use it, but it can expose an alias, because eventually the resampling actor could have other parameters in the request (I was thinking that maybe, just **maybe**, we could want at some point to request a metric to be resampled with a particular resampling function after we talked about resampling a `xxx_max` metric for example). In that case it will have to define it's own request.

All that said, it is still a style issue I guess, so I won't oppose if you want to leave it like this. :)

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (leandro-lucarella-frequenz) [reply src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: OK, I forgot to mention but I think the PR is fine without this too, it was more of a thought and to check if I was understanding correctly, but if it is a low hanging fruit and want to add it now I won't oppose either :)

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (leandro-lucarella-frequenz) [reply src/frequenz/sdk/actor/data_sourcing/common.py]: Closing so we have the discussion only in one place: https://github.com/frequenz-floss/frequenz-sdk-python/pull/52#discussion_r1005486650

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (leandro-lucarella-frequenz) [reply src/frequenz/sdk/actor/data_sourcing/common.py]: Closing so we have all the discussion about where to put these types in one place: https://github.com/frequenz-floss/frequenz-sdk-python/pull/52#discussion_r1005486650

File: src/frequenz/sdk/actor/data_sourcing/microgrid_api_source.py
Comment (sahas-subramanian-frequenz) [reply src/frequenz/sdk/actor/data_sourcing/microgrid_api_source.py]: This is one of the stupid things about python that lead to confusing bugs for me.  If the language always insists on `.keys()`,  that would have been amazing.  But I will remove it, to be consistent.

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (shsms) [reply src/frequenz/sdk/actor/data_sourcing/common.py]: The `DataSourcingActor` will also be the first source of `Sample`s.  What do you think about just `frequenz/sdk/actor/data_sourcing/types.py` for all three?

(Also I had a typo in the path I mentioned earlier, that's now fixed)

File: src/frequenz/sdk/actor/data_sourcing/data_sourcing.py
Comment (sahas-subramanian-frequenz) [inline src/frequenz/sdk/actor/data_sourcing/data_sourcing.py:26]: we probably need a global singleton registry as well,  before this PR can be merged.

Comment (leandro-lucarella-frequenz) [reply src/frequenz/sdk/actor/data_sourcing/data_sourcing.py:26]: Why? I mean why it is a requirement for this to be merged?

Comment (sahas-subramanian-frequenz) [reply src/frequenz/sdk/actor/data_sourcing/data_sourcing.py:26]: Because we're passing the registry as an argument to an actor,  and that's not ideal.

We can of course merge,  but then we'll have to come back and fix this afterwards.

File: src/frequenz/sdk/actor/data_sourcing/data_sourcing.py
Comment (sahas-subramanian-frequenz) [reply src/frequenz/sdk/actor/data_sourcing/data_sourcing.py]: done

File: src/frequenz/sdk/actor/data_sourcing/microgrid_source.py
Comment (sahas-subramanian-frequenz) [reply src/frequenz/sdk/actor/data_sourcing/microgrid_source.py]: this is done for now.  Reusing channels will get complicated when we start adding parquet as a source,  but we can deal that then.

File: src/frequenz/sdk/actor/data_sourcing/microgrid_api_source.py
Comment (sahas-subramanian-frequenz) [reply src/frequenz/sdk/actor/data_sourcing/microgrid_api_source.py]: done

File: src/frequenz/sdk/actor/data_sourcing/common.py
Comment (leandro-lucarella-frequenz) [reply src/frequenz/sdk/actor/data_sourcing/common.py]: Haha, I thought about `Sample` but since I didn't have a good case for it I left it out in the hope that you won't notice :trollface: 

OK, now that you forced me to think about it harder, I found out why `Sample` is different for me: If we ever make the resampler agnostic of the data sourcing actor, `Sample` will be a more general concept, where other sourcing actors will also use `Sample`, but not `ComponentMetricId` and `ComponentMetricRequest`. This is why I think `Sample` is a more general data pipeline concept, not necessarily tied to the data sourcing actor, while the other too I don't see them being used outside of it (independently I mean, not to communicate with the data sourcing actor).

But since this is an hypothetical case for now, I would be fine to add it too to the data sourcing actor. At the end I think we'll probably need to re-arrange a thing of two as we advance more with the data pipeline and the structure is more clear.

About `types`, I'm not crazy about general modules like `common` and `types` but `types` is definitely better as it is more scoped than `common`. So personally I would still have `ComponentMetricId` and `ComponentMetricRequest` very close to the class that gives live to those classes (again, for me `Sample` is different), but `types` is fine too.

**Review by leandro-lucarella-frequenz — COMMENTED** [review]
LGTM, I can approve after the fixups are squashed (otherwise I will have to approve again anyway :)

**Review by leandro-lucarella-frequenz — APPROVED** [review]

**Pull request merged**
